x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/enrich_signal_threat_matches.ts

```diff
@@ -0,0 +1,113 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get, isObject } from 'lodash';
+
+import type { SignalSearchResponse, SignalSourceHit } from '../types';
+import type {
+  GetMatchedThreats,
+  ThreatIndicator,
+  ThreatListItem,
+  ThreatMatchNamedQuery,
+} from './types';
+import { extractNamedQueries } from './utils';
+
+const DEFAULT_INDICATOR_PATH = 'threat.indicator';
+const getSignalId = (signal: SignalSourceHit): string => signal._id;
+
+export const groupAndMergeSignalMatches = (signalHits: SignalSourceHit[]): SignalSourceHit[] => {
+  const dedupedHitsMap = signalHits.reduce<Map<unknown, SignalSourceHit>>((acc, signalHit) => {
+    const signalId = getSignalId(signalHit);
+    if (!acc.has(signalId)) {
+      acc.set(signalId, signalHit);
+    } else {
+      const existingSignalHit = acc.get(signalId) as SignalSourceHit;
+      const existingQueries = existingSignalHit?.matched_queries ?? [];
+      const newQueries = signalHit.matched_queries ?? [];
+      existingSignalHit.matched_queries = [...existingQueries, ...newQueries];
+
+      acc.set(signalId, existingSignalHit);
+    }
+
+    return acc;
+  }, new Map());
+  const dedupedHits = Array.from(dedupedHitsMap.values());
+  return dedupedHits;
+};
+
+export const buildMatchedIndicator = ({
+  queries,
+  threats,
+  indicatorPath = DEFAULT_INDICATOR_PATH,
+}: {
+  queries: ThreatMatchNamedQuery[];
+  threats: ThreatListItem[];
+  indicatorPath?: string;
+}): ThreatIndicator[] =>
+  queries.map((query) => {
+    const matchedThreat = threats.find((threat) => threat._id === query.id);
+    const indicatorValue = get(matchedThreat?._source, indicatorPath) as unknown;
+    const indicator = [indicatorValue].flat()[0] ?? {};
+    if (!isObject(indicator)) {
+      throw new Error(`Expected indicator field to be an object, but found: ${indicator}`);
+    }
+    const atomic = get(matchedThreat?._source, query.value) as unknown;
+    const type = get(indicator, 'type') as unknown;
+
+    return {
+      ...indicator,
+      matched: { atomic, field: query.field, type },
+    };
+  });
+
+export const enrichSignalThreatMatches = async (
+  signals: SignalSearchResponse,
+  getMatchedThreats: GetMatchedThreats
+): Promise<SignalSearchResponse> => {
+  const signalHits = signals.hits.hits;
+  if (signalHits.length === 0) {
+    return signals;
+  }
+
+  const uniqueHits = groupAndMergeSignalMatches(signalHits);
+  const signalMatches = uniqueHits.map((signalHit) => extractNamedQueries(signalHit));
+  const matchedThreatIds = [...new Set(signalMatches.flat().map(({ id }) => id))];
+  const matchedThreats = await getMatchedThreats(matchedThreatIds);
+  const matchedIndicators = signalMatches.map((queries) =>
+    buildMatchedIndicator({ queries, threats: matchedThreats })
+  );
+
+  const enrichedSignals = uniqueHits.map((signalHit, i) => {
```

Comment:
We should type this as

`const enrichedSignals: SignalSourceHit[] ...`

so that https://github.com/elastic/kibana/pull/89899/files#diff-23d7709dfeb5dfa27a93e789abe329fef4934d9f3b0c78850045e22e34f1f7fcR104 have equivalent types.

Reply:
https://github.com/elastic/kibana/pull/89899/commits/9b8e343e1b9b5b13a30fb916bfc0d642e99fccc8. I noted in that comment that I _believe_ this just affects where a type mismatch error is raised, but earlier is better 👍